[Merged by Bors] - feat: smooth embeddings

grunweg opened the pull request:
Define smooth embeddings between smooth manifolds: a map is a smooth embedding iff it is both a smooth immersion and a topological embedding. We also prove basic properties (such as being stable under products); stronger results will be added in future PRs.

Note that unlike smooth immersions, we do not add definitions for
- being a smooth embedding on a set (as this doesn't exist in the topological category either),
- being an embedding at a point (being an embedding is a global notion; moreover, every immersion is locally an embedding anyway, so "being an immersion at x" is sufficient to state this)
- variants with respect to a fixed complement: the motivation for immersions (considering the equivalent local descriptions of smooth submanifolds) do not apply to this global notion.

Along the way, we add that the identity map is a smooth immersion, and rename the `ofOpens` declaration to `of_opens` (which would have been the correct name all along). As they were added just two days ago, we omit a deprecation.

-----

- [x] depends on: #28853

---
<!-- Your PR title will become the first line of the commit message.

In this box, the text above the `---` (if not empty) will be appended
to the commit message, and can be used to give additional context or
details. Please leave a blank newline before the `---`, otherwise GitHub
will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
feat: smooth embeddings

Define smooth embeddings between smooth manifolds: a map is a smooth embedding iff it is both
a smooth immersion and a topological embedding. We also prove basic properties (such as congruence
or being stable under products); stronger results will come in future PRs.

Note that unlike smooth immersions, we do not add definitions for
- being a smooth embedding on a set (as this doesn't exist in the topological category either),
- being an embedding at a point (being an embedding is a global notion; moreover, every immersion is
  locally an embedding anyway, so "being an immersion at x" is sufficient to state this)
- variants with respect to a fixed complement: the motivation for immersions (considering the
  equivalent local descriptions of smooth submanifolds) do not apply to this global notion.

grunweg committed:
Review comments

chrisflav commented on Mathlib/Geometry/Manifold/SmoothEmbedding.lean:
```diff
@@ -0,0 +1,87 @@
+@[mk_iff]
+structure IsSmoothEmbedding (f : M → N) where
+  isImmersion : IsImmersion I J n f
+  isImbedding : IsEmbedding f
```
```suggestion
  isEmbedding : IsEmbedding f
```

grunweg committed:
Update Mathlib/Geometry/Manifold/SmoothEmbedding.lean

Co-authored-by: Christian Merten <christian@merten.dev>

grunweg committed:
Remove congr lemma; add TODOs, other review comments

peabrainiac reviewed (approved):
Other than one small comment this looks good to me - thank you again for all the work you put into this chain of PRs! I am very happy to finally see immersions and embeddings make their way into mathlib.

Out of curiosity, how far are you from establishing the universal property of embeddings ($f$ is an embedding iff for all $g$, $f\circ g$ is smooth iff $g$ is) and the fact that e.g. the inclusion of an open subset is an open smooth embedding? These seem like two of the most important basic facts about embeddings to me, next to identities and compositions of embeddings being embeddings.

grunweg committed:
Update Mathlib/Geometry/Manifold/SmoothEmbedding.lean

Co-authored-by: Ben Eltschig <43812953+peabrainiac@users.noreply.github.com>

grunweg commented:
"The inclusion of an open subset is a smooth embedding" should be easy. ("Open smooth embedding" would be easy if I add that definition; as I'm not planning on adding lots of API on that, I wasn't going to add that. Be my guest!) #32760 tries to do the embedding half.

grunweg commented:
#28865 does the first step towards the universal property: it proves half of it, for immersions. (Upgrading that to embeddings should be easy; the additional piece is in mathlib IIRC.) It's blocked on proving e.g. smoothness.

Thanks for the kind words. It's great to hear that this matters to others also!

chrisflav reviewed (approved):
Thanks!
maintainer delegate

grunweg committed:
Update Mathlib/Geometry/Manifold/SmoothEmbedding.lean

Co-authored-by: Christian Merten <christian@merten.dev>

grunweg committed:
Merge branch 'master' into smoothembeddings-products

grunweg committed:
chore: rename IsImmersion.ofOpen to of_opens

grunweg committed:
feat: add IsSmoothEmbedding.of_opens

ocfnash commented:
It would be nice if we could include some trivial examples of some embeddings. E.g., is this true:
```lean
example (f : M ≃ₜ N) (hf : ContMDiff I J n f) : IsSmoothEmbedding I J n f := sorry
```
or can we show a diffeomorphism is an embedding?

grunweg commented:
A smooth map which is a topological embedding need not be an immersion; my standard example is https://math.stackexchange.com/a/2583667. That doesn't cover your question precisely, but the second answer https://math.stackexchange.com/a/3769328  can be made to. Let me clarify the module doc-string accordingly.

A diffeomorphism is certainly a smooth embedding; let me add that.

grunweg committed:
Add TODO about diffeomorphisms

grunweg commented:
I spoke too soon: mathematically, it's indeed easy to prove that a diffeomorphism is a smooth embedding, but not so with the current definition. The conceptually correct result is that a local diffeomorphism is a smooth immersion: doing so requires proving that injective differential (resp. split differential in infinite dimension) implies being an immersion. This will come in future PRs (and requires the inverse function theorem).
You *could* cheat around this by constructing pullback charts using a diffeomorphism; this is code I'd delete again once the proper definition is merged. I'm not convinced this is worth it. I have added a TODO and will prove that the identity is a smooth embedding, as at least a very basic test.

grunweg committed:
State the result I want; quote explicit counterexamples

grunweg committed:
feat: the identity map is a smooth embedding

grunweg commented:
Added that the identity map is a smooth embedding. (Should this be protected?)

grunweg commented:
-awaiting-author

ocfnash reviewed (commented):
OK, it's a little concerning that we cannot easily construct examples even from diffeomorphisms but I do appreciate your point about not wanting to write throw-away code.

Thanks!

bors d+

grunweg committed:
Protect .id; better naming scheme

grunweg commented on Mathlib/Geometry/Manifold/SmoothEmbedding.lean:
```diff
@@ -0,0 +1,109 @@
+namespace Manifold
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {E E' E'' E''' F F' : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
```
Good points, both of them. Thank you!

grunweg committed:
Add to overview
